### PR TITLE
rustdoc: Rename `@has FILE PATTERN` to `@hasraw FILE PATTERN`

### LINKED ISSUE #100354: rustdoc: Rename `@has FILE PATTERN` to `@hastext`
`@has FILE PATTERN` treats `PATTERN` as literal text, while `@has FILE PATTERN ANOTHER_PATTERN` treats `PATTERN` as an XPath (and `ANOTHER_PATTERN` as literal text). This is quite confusing and left me perplexed at why my test was failing. IMO we should have a different name for these two versions. The literal (arity-2) version could be named `@hastext`.

This would also rename the equivalent for `@matches` to `@matchestext`.

_Originally posted by @camelid in https://github.com/rust-lang/rust/issues/100348#issuecomment-1210145559_

### PR DESCRIPTION
**Pull request by camelid**
Fixes #100354.

### CODE REVIEW

**Comment by camelid** [comment]
For future reference, I used variations of this Find&Replace regex:
```regex
@matches ([^ ]+) "([^"]+)"$
```
```regex
@matchestext $1 $2
```

**Review by camelid — COMMENTED** [review]
There are several instances of multiline commands that were incorrectly that I'll need to fixup manually.

**Comment by camelid** [comment]
Oops forgot to update `@!has`.

File: src/test/rustdoc/internal.rs [hunk 1]
```diff
@@ -2,14 +2,14 @@
...
 //      '//*[@class="item-right docblock-short"]/span[@class="stab unstable"]'
-// @!matches internal/index.html \
+// @!matchesraw internal/index.html \
 //      '//*[@class="item-right docblock-short"]/span[@class="stab internal"]'
 // @matches - '//*[@class="item-right docblock-short"]' 'Docs'
 
-// @!has internal/struct.S.html '//*[@class="stab unstable"]'
-// @!has internal/struct.S.html '//*[@class="stab internal"]'
+// @!hasraw internal/struct.S.html '//*[@class="stab unstable"]'
+// @!hasraw internal/struct.S.html '//*[@class="stab internal"]'
```
Comment (camelid) [inline src/test/rustdoc/internal.rs:12]: These were all using `@!has`/`@!matches` incorrectly and need to be fixed.

File: src/test/rustdoc/issue-61592.rs [hunk 2]
```diff
@@ -5,11 +5,11 @@ extern crate foo;
...
 // @has - '//code' 'pub use foo::FooTrait as _;'
-// @!has - '//a[@href="trait._.html"]'
+// @!hasraw - '//a[@href="trait._.html"]'
 pub use foo::FooTrait as _;
 
 // @has issue_61592/index.html
 // @has - '//a[@href="#reexports"]' 'Re-exports'
 // @has - '//code' 'pub use foo::FooStruct as _;'
-// @!has - '//a[@href="struct._.html"]'
+// @!hasraw - '//a[@href="struct._.html"]'
```
Comment (camelid) [inline src/test/rustdoc/issue-61592.rs:14]: More incorrect usage.

File: src/test/rustdoc/no-crate-filter.rs [hunk 3]
```diff
@@ -2,5 +2,5 @@
 
 // compile-flags: -Z unstable-options --disable-per-crate-search
 
-// @!has 'foo/struct.Foo.html' '//*[id="crate-search"]'
+// @!hasraw 'foo/struct.Foo.html' '//*[id="crate-search"]'
```
Comment (camelid) [inline src/test/rustdoc/no-crate-filter.rs:5]: Ditto.

File: src/test/rustdoc/recursive-deref.rs [hunk 4]
```diff
@@ -101,7 +101,7 @@ impl I {
 }
 
 // @has recursive_deref/struct.H.html '//h3[@class="code-header in-band"]' 'impl Deref for H'
-// @!has '-' '//*[@id="deref-methods-I"]'
+// @!hasraw '-' '//*[@id="deref-methods-I"]'
```
Comment (camelid) [inline src/test/rustdoc/recursive-deref.rs:104]: Ditto.

File: src/test/rustdoc/remove-url-from-headings.rs [hunk 5]
```diff
@@ -1,7 +1,7 @@
 #![crate_name = "foo"]
 
 // @has foo/fn.foo.html
-// @!has - '//a[@href="http://a.a"]'
+// @!hasraw - '//a[@href="http://a.a"]'
```
Comment (camelid) [inline src/test/rustdoc/remove-url-from-headings.rs:4]: Ditto.

File: src/test/rustdoc/sized_trait.rs [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
 // @has foo/struct.Bar.html
-// @!has - '//*[@id="impl-Sized"]'
+// @!hasraw - '//*[@id="impl-Sized"]'
 pub struct Bar {
     a: u16,
 }
 
 // @has foo/struct.Foo.html
-// @!has - '//*[@id="impl-Sized"]'
+// @!hasraw - '//*[@id="impl-Sized"]'
```
Comment (camelid) [inline src/test/rustdoc/sized_trait.rs:10]: More.

File: src/test/rustdoc/table-in-docblock.rs [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 // @has foo/struct.Foo.html
 // @count - '//*[@class="docblock"]/div/table' 2
-// @!has - '//*[@class="docblock"]/table'
+// @!hasraw - '//*[@class="docblock"]/table'
```
Comment (camelid) [inline src/test/rustdoc/table-in-docblock.rs:5]: More.

File: src/test/rustdoc/toggle-item-contents.rs [hunk 8]
```diff
@@ -72,7 +72,7 @@ pub enum Enum {
 }
 
 // @has 'toggle_item_contents/enum.EnumStructVariant.html'
-// @!has - '//details[@class="rustdoc-toggle type-contents-toggle"]'
+// @!hasraw - '//details[@class="rustdoc-toggle type-contents-toggle"]'
```
Comment (camelid) [inline src/test/rustdoc/toggle-item-contents.rs:75]: More.

File: src/test/rustdoc/trait-impl-items-links-and-anchors.rs [hunk 9]
```diff
@@ -59,7 +59,7 @@ pub struct MyStruct;
 
 // We check that associated items with default values aren't generated in the implementors list.
 impl MyTrait for (u8, u8) {
-    // @!has trait_impl_items_links_and_anchors/trait.MyTrait.html '//div[@id="associatedconstant.VALUE-4"]'
+    // @!hasraw trait_impl_items_links_and_anchors/trait.MyTrait.html '//div[@id="associatedconstant.VALUE-4"]'
```
Comment (camelid) [inline src/test/rustdoc/trait-impl-items-links-and-anchors.rs:62]: More.

File: src/test/rustdoc/trait-impl.rs [hunk 10]
```diff
@@ -40,7 +40,7 @@ impl Trait for Struct {
     fn c() {}
 
     // @has - '//*[@id="method.d"]/../../div[@class="docblock"]/p' 'Escaped formatting a*b*c* works'
-    // @!has - '//*[@id="method.d"]/../../div[@class="docblock"]/p/em'
+    // @!hasraw - '//*[@id="method.d"]/../../div[@class="docblock"]/p/em'
```
Comment (camelid) [inline src/test/rustdoc/trait-impl.rs:43]: And more...

File: src/test/rustdoc/tuple-struct-fields-doc.rs [hunk 11]
```diff
@@ -5,7 +5,7 @@
 // @has - '//h3[@class="sidebar-title"]/a[@href="#fields"]' 'Tuple Fields'
 // @has - '//*[@id="structfield.0"]' '0: u32'
 // @has - '//*[@id="main-content"]/div[@class="docblock"]' 'hello'
-// @!has - '//*[@id="structfield.1"]'
+// @!hasraw - '//*[@id="structfield.1"]'
```
Comment (camelid) [inline src/test/rustdoc/tuple-struct-fields-doc.rs:8]: Yet another.

**Review by camelid — COMMENTED** [review]
Left comments at each place where I noticed that `@!has` was being used incorrectly before. It used `@!has` (or the other commands) with an XPath that was actually treated as a literal string of HTML code. All or almost all of the uses were with the negated commands (e.g., `@!has) since otherwise the tests would've noticeably failed.

**Comment by camelid** [comment]
Ok, I fixed all the incorrect uses of `@!has`, and the tests passed. This is ready for review!

**Comment by camelid** [comment]
Maybe r? @GuillaumeGomez 

(@Mark-Simulacrum feel free to review if you'd like, though not sure why highfive assigned you since this is rustdoc-related)

**Comment by GuillaumeGomez** [comment]
I'm confused: didn't you say in the issue that the problem was the commands should have a different name when then they had 2 vs 3 arguments? Because otherwise, I don't see the point of this PR: the command name is longer and definitely not more helpful.

**Comment by camelid** [comment]
That's what this PR does: it renames the 2-argument version to `@hasraw`. The 3-argument version is still `@has`. So this addresses the issue: they now have different names.

As you can see in the last commit, it turns out that a lot of tests made the same mistake I did and thought `@has FILE PATTERN` was the same as `@has FILE PATTERN ''` (which is what they intended).

Does that help? If not, please let me know which part you want me to clarify.

**Comment by camelid** [comment]
Note the 3-argument version keeps its name in this PR; only the 2-argument version is different.

**Comment by GuillaumeGomez** [comment]
Sorry, I completely missed it somehow. I'll make another review shortly then. Thanks for the extra explanation!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors r+

**Pull request merged**